pytential/qbx/__init__.py

```diff
@@ -87,111 +90,135 @@ class QBXLayerPotentialSource(LayerPotentialSourceBase):
-                    "target_association_tolerance. "
-                    "Using target_stick_out_factor is deprecated "
-                    "and will stop working in 2018.",
-                    DeprecationWarning, stacklevel=2)
-
-            if target_association_tolerance is not _not_provided:
-                raise TypeError("May not pass both target_association_tolerance and "
-                        "target_stick_out_factor.")
-
-            target_association_tolerance = target_stick_out_factor
-
-        del target_stick_out_factor
+            raise ValueError("'qbx_order' must be provided.")
+        assert isinstance(qbx_order, int)
 
         if target_association_tolerance is _not_provided:
-            target_association_tolerance = float(
-                    np.finfo(density_discr.real_dtype).eps) * 1e3
+            target_association_tolerance = (
+                1.0e+3 * float(np.finfo(density_discr.real_dtype).eps))
+        assert isinstance(target_association_tolerance, float)
 
         if fmm_order is not None and fmm_level_to_order is not None:
-            raise TypeError("may not specify both fmm_order and fmm_level_to_order")
+            raise TypeError(
+                "may not specify both 'fmm_order' and 'fmm_level_to_order'.")
 
         if _box_extent_norm is None:
             _box_extent_norm = "l2"
+        assert isinstance(_box_extent_norm, str)
 
         if _from_sep_smaller_crit is None:
             # This seems to win no matter what the box extent norm is
             # https://gitlab.tiker.net/papers/2017-qbx-fmm-3d/issues/10
             _from_sep_smaller_crit = "precise_linf"
+        assert isinstance(_from_sep_smaller_crit, str)
 
         if fmm_level_to_order is None:
-            if fmm_order is False:
+            if fmm_order is False or fmm_order is None:
                 fmm_level_to_order = False
             else:
+                assert isinstance(fmm_order, int) and not isinstance(fmm_order, bool)
+
                 def fmm_level_to_order(kernel, kernel_args, tree, level):  # noqa pylint:disable=function-redefined
                     return fmm_order
+        assert isinstance(fmm_level_to_order, bool) or callable(fmm_level_to_order)
 
         if _max_leaf_refine_weight is None:
             if density_discr.ambient_dim == 2:
                 # FIXME: This should be verified now that l^2 is the default.
                 _max_leaf_refine_weight = 64
             elif density_discr.ambient_dim == 3:
-                # For static_linf/linf: https://gitlab.tiker.net/papers/2017-qbx-fmm-3d/issues/8#note_25009  # noqa
-                # For static_l2/l2: https://gitlab.tiker.net/papers/2017-qbx-fmm-3d/issues/12  # noqa
+                # For static_linf/linf:
+                #   https://gitlab.tiker.net/papers/2017-qbx-fmm-3d/issues/8#note_25009
+                # For static_l2/l2:
+                #   https://gitlab.tiker.net/papers/2017-qbx-fmm-3d/issues/12
```

Comment:
These links don't seem to be accessible from the outside?

Reply:
That's fair! I'm not advocating to necessarily make them public or anything, so added a little note in https://github.com/inducer/pytential/compare/75d27feb90f28eab9ac559dd66c14cb25383f007..5349918da7ec88001f8a9f6dbf9fa69a3528747e